File: node/src/facilitators/aifinpay.ts
import nacl from "tweetnacl";
import bs58 from "bs58";
import type { Agent } from "../agent.js";
import { sha256 } from "../crypto.js";
import type {
  AuthPayload,
  AuthRequestContext,
  Facilitator,
  PayOptions,
} from "./base.js";

/**
 * Native AiFinPay flavor.
 *
 * Wire format:
 *   - 402 carries a JSON body with `protocol: "AiFinPay vX"` and
 *     `agreement_hash`, `treasury_vault`, `x-nonce` …
 *   - Client retries with three headers:
 *       x-agent-pubkey, x-nonce, x-signature
 *   - Signature: Ed25519 over SHA-256(canonical v2 request binding)
 */
export class AiFinPayFacilitator implements Facilitator {
  static readonly name = "aifinpay";
  readonly name = "aifinpay";

  static async detect(resp: Response): Promise<boolean> {
    if (resp.status !== 402) return false;
    let body: unknown;
    try {
      body = await resp.clone().json();
    } catch {
      return false;
    }
    if (typeof body !== "object" || body === null) return false;
    const b = body as Record<string, unknown>;
    if (typeof b.protocol === "string" && b.protocol.startsWith("AiFinPay")) {
      return true;
    }
    // Fallback fingerprint when an upstream proxy strips `protocol`.
    return (
      ("agreement_hash" in b || "manifesto" in b) &&
      ("treasury_vault" in b || "program_id" in b)
    );
  }

  async buildAuth(
    resp: Response,
    agent: Agent,
    _opts: PayOptions,
    context?: AuthRequestContext,
  ): Promise<AuthPayload> {
    if (!context) {
      throw new Error(
        "AiFinPay native auth v1 is no longer supported. Retry through Agent.pay() so the SDK can bind the challenge to the request.",
      );
    }
    if (context.url.origin !== context.trustedOrigin) {
      throw new Error(
        `refusing native authentication for untrusted origin ${context.url.origin}; configure Agent.baseUrl for that facilitator explicitly`,
      );
    }
    if (resp.url && new URL(resp.url).origin !== context.trustedOrigin) {
      throw new Error("refusing native authentication after a cross-origin response");
    }

    const challenge = await this.inbandChallenge(resp, context.bodyDigest);
    if (!challenge) {
      throw new Error(
        "AiFinPay native auth v2 requires an in-band request-bound challenge. Retry the original request without credentials.",
      );
    }
    const message = JSON.stringify([
      "AiFinPay-x402",
      "v2",
      challenge.nonce,
      agent.address,
      context.trustedOrigin,
      context.method.toUpperCase(),
      `${context.url.pathname}${context.url.search}`,
      context.bodyDigest,
      challenge.expiresAt,
    ]);
    const msg = new TextEncoder().encode(message);
    const digest = await sha256(msg);
    const sig = nacl.sign.detached(digest, agent.secretKey);
    return {
      headers: {
        "x-agent-pubkey": agent.address,
        "x-nonce": challenge.nonce,
        "x-signature": bs58.encode(sig),
        "x-aifinpay-auth-version": "2",
      },
    };
  }

  private async inbandChallenge(
    resp: Response,
    expectedBodyDigest: string,
  ): Promise<{ nonce: string; expiresAt: number; bodyDigest: string } | null> {
    let body: unknown;
    try {
      body = await resp.clone().json();
    } catch {
      return null;
    }
    if (typeof body !== "object" || body === null) return null;
    const b = body as Record<string, unknown>;
    const nonce = b["x-nonce"];
    const expiresAt = b["x-nonce-expires-at"];
    const version = b["x-aifinpay-auth-version"];
    const bodyDigest = b["x-aifinpay-body-sha256"];
    if (
      typeof nonce !== "string" ||
      !/^[A-Za-z0-9_-]{16,128}$/.test(nonce) ||
      (version !== 2 && version !== "2") ||
      (typeof expiresAt !== "string" && typeof expiresAt !== "number") ||
      !/^[1-9][0-9]{0,15}$/.test(String(expiresAt)) ||
      typeof bodyDigest !== "string" ||
      !/^[0-9a-f]{64}$/.test(bodyDigest) ||
      bodyDigest !== expectedBodyDigest
    ) {
      return null;
    }
    const parsedExpiry = Number(expiresAt);
    const now = Date.now();
    return Number.isSafeInteger(parsedExpiry) && parsedExpiry > now && parsedExpiry <= now + 5 * 60_000
      ? { nonce, expiresAt: parsedExpiry, bodyDigest }
      : null;
  }
}

File: python/aifinpay/facilitators/aifinpay.py
"""Native AiFinPay flavor — three custom headers, JSON body in the 402."""
from __future__ import annotations

import hashlib
import json
import re
import time
from typing import TYPE_CHECKING, Any, Optional
from urllib.parse import urlsplit

import base58
import requests

from .base import Facilitator, PayOptions, canonical_origin

if TYPE_CHECKING:
    from ..client import Agent


class AiFinPayFacilitator:
    """Adapter for the AiFinPay native x402 flow.

    Wire format:
        - 402 carries a JSON body with `program_id`, `manifesto`,
          `agreement_hash`, `treasury_vault`, …
        - Client retries with three headers:
            x-agent-pubkey, x-nonce, x-signature
        - Signature: Ed25519 over SHA-256(canonical v2 request binding)
    """

    name = "aifinpay"

    @staticmethod
    def detect(resp: requests.Response) -> bool:
        if resp.status_code != 402:
            return False
        try:
            body = resp.json()
        except ValueError:
            return False
        if not isinstance(body, dict):
            return False
        # AiFinPay 402 carries `protocol: "AiFinPay vX.Y"` plus either
        # `agreement_hash` (most common) or `manifesto` ref.
        protocol = body.get("protocol", "")
        if isinstance(protocol, str) and protocol.startswith("AiFinPay"):
            return True
        # Fallback fingerprint when an upstream proxy strips `protocol`.
        return ("agreement_hash" in body or "manifesto" in body) and (
            "treasury_vault" in body or "program_id" in body
        )

    @staticmethod
    def _sign_request(
        agent: "Agent",
        nonce: str,
        origin: str,
        method: str,
        resource: str,
        body_digest: str,
        expires_at: int,
    ) -> str:
        # Must match JSON.stringify([...]) in the Node SDK and backend.
        msg = json.dumps(
            [
                "AiFinPay-x402",
                "v2",
                nonce,
                agent.address,
                origin,
                method.upper(),
                resource,
                body_digest,
                expires_at,
            ],
            separators=(",", ":"),
            ensure_ascii=True,
        ).encode()
        digest = hashlib.sha256(msg).digest()
        sig = agent._sk.sign(digest).signature
        return base58.b58encode(sig).decode()

    def build_auth(
        self,
        resp: requests.Response,
        agent: "Agent",
        opts: PayOptions,
        context: Optional[dict[str, Any]] = None,
    ) -> dict:
        if not context:
            raise ValueError(
                "AiFinPay native auth v1 is no longer supported. Retry through Agent.pay() so the SDK can bind the challenge to the request."
            )
        target = urlsplit(str(context["url"]))
        trusted_origin = str(context["trusted_origin"])
        target_origin = canonical_origin(str(context["url"]))
        if target_origin != trusted_origin:
            raise ValueError(
                f"refusing native authentication for untrusted origin {target_origin}; configure Agent.base_url for that facilitator explicitly"
            )
        if resp.url:
            try:
                response_origin = canonical_origin(resp.url)
            except ValueError as exc:
                raise ValueError("refusing native authentication for an invalid response origin") from exc
            if response_origin != trusted_origin:
                raise ValueError("refusing native authentication after a cross-origin response")

        body_digest = str(context["body_digest"])
        challenge = self._inband_challenge(resp, body_digest)
        if challenge is None:
            raise ValueError(
                "AiFinPay native auth v2 requires an in-band request-bound challenge. Retry the original request without credentials."
            )
        nonce, expires_at = challenge
        resource = target.path or "/"
        if target.query:
            resource += f"?{target.query}"
        headers = {
            "x-agent-pubkey": agent.address,
            "x-nonce": nonce,
            "x-signature": self._sign_request(
                agent,
                nonce,
                trusted_origin,
                str(context["method"]),
                resource,
                body_digest,
                expires_at,
            ),
            "x-aifinpay-auth-version": "2",
        }
        return {"headers": headers}

    @staticmethod
    def _inband_challenge(
        resp: requests.Response, expected_body_digest: str = ""
    ) -> tuple[str, int] | None:
        """Read the v2 challenge issued for this exact unauthenticated request."""
        try:
            body = resp.json()
        except ValueError:
            return None
        if not isinstance(body, dict):
            return None
        nonce = body.get("x-nonce")
        expires_at = body.get("x-nonce-expires-at")
        version = body.get("x-aifinpay-auth-version")
        body_digest = body.get("x-aifinpay-body-sha256")
        if (
            not isinstance(nonce, str)
            or re.fullmatch(r"[A-Za-z0-9_-]{16,128}", nonce) is None
            or str(version) != "2"
            or not isinstance(expires_at, (str, int))
            or not (str(expires_at).isdigit() and 1 <= len(str(expires_at)) <= 16)
            or not isinstance(body_digest, str)
            or len(body_digest) != 64
            or any(ch not in "0123456789abcdef" for ch in body_digest)
            or body_digest != expected_body_digest
        ):
            return None
        parsed_expiry = int(expires_at)
        now = int(time.time() * 1000)
        return (
            (nonce, parsed_expiry)
            if now < parsed_expiry <= now + 5 * 60_000
            else None
        )
